app/restore_to_file.go

```diff
@@ -59,6 +61,53 @@ func RestoreToFileCmd() cli.Command {
 	}
 }
 
+type Restore struct {
+	sync.RWMutex
+	SnapshotName    string
+	RestoreProgress int
+	RestoreError    error
+}
+
+func (r *Restore) UpdateRestoreStatus(snapshot string, rp int, re error) {
+	r.Lock()
+	defer r.Unlock()
+
+	logrus.Errorf("Updating progress[%v], snapshot[%v], error[%v]", rp, snapshot, re)
+	r.SnapshotName = snapshot
+	r.RestoreProgress = rp
+	r.RestoreError = re
+}
+
+func checkRestore(backupURL string) error {
+	restoreObj := &Restore{}
+
+	config := &backupstore.DeltaRestoreConfig{
+		BackupURL:    backupURL,
+		DeltaOps:     restoreObj,
+		SnapshotName: BackupFilePath,
+	}
+
+	err := backupstore.RestoreDeltaBlockBackup(config)
+	if err != nil {
+		return err
+	}
+
+	for j := 0; j < 600; j++ {
```

Comment:
Same here as in the previous case.

Reply:
Restore-to-file calls backupstore.RestoreDeltaBlockBackup(config) directly. I tried to make it asynchronous but when the restore is happening in the background, the process that initiated it terminates. 
```
$ docker exec 1d8d057cb2b8 longhorn --url http://localhost:9501 backup restore-to-file vfs:///data/backupstore?backup=backup-f571a76c907d4463\u0026volume=vol-name
time="2019-06-27T00:19:31Z" level=info msg="Running restore to file command: backup-url=vfs:///data/backupstore?backup=backup-f571a76c907d4463u0026volume=vol-name backing-file= output-file= output-format=qcow2"
time="2019-06-27T00:19:31Z" level=info msg="Output file path=/volume.qcow2"
time="2019-06-27T00:19:31Z" level=info msg="Start to restore vfs:///data/backupstore?backup=backup-f571a76c907d4463u0026volume=vol-name to backup.img"
time="2019-06-27T00:19:31Z" level=info msg="Done running restore to file command. Produced image: "
```

Do we have to keep it waiting, in that case, why not keep this a blocking operation?

Reply:
The restore to file is a blocking operation, but we need to show the user the progress and don't timeout based on the time only. We need to timeout based on the progress as in the other cases.

We can use e.g. https://github.com/cheggaaa/pb to show the progress bar in the console.

Reply:
This progress bar package seems to have lots of issues. Couldn't compile it as the package has many dependency errors. Will just use the timer logic in the restore_to_file.

Reply:
You need to add those dependency in the vendor.conf and regenerate dependency use trash.

Reply:
and use v1 for now. v2 is in beta.